### ISSUE TITLE
Extend Alarmo MQTT to include Notifications (Feature Request)

### ISSUE THREAD

**Issue by thanksmister** [nielsfaber/alarmo#92]
Hello fellow Home Assistant Alarm creator.  I am the creator and maintainer of the MQTT Alarm Panel application for Home Assistant.  MQTT Alarm Panel is an Android application for wall mounted tablets that works with the MQTT Manual component in HA.  My application already uses MQTT Manual but I have also added the missing night mode (and eventually custom bypass), ability to send a code/command payload, and sensors to display the state of a door sensor for example (open or closed)

Here is my [Github repository](https://github.com/thanksmister/android-mqtt-alarm-panel) and the application is also available from [Google Play store](https://play.google.com/store/apps/details?id=com.thanksmister.iot.mqtt.alarmpanel). 

Unfortunately, the MQTT Manual component has been neglected for some time and our PR to update the component was rejected.   So I am reaching out to you because I had some suggestions four how we might extend the MQTT capabilities of your Alarmo integration that would be great for MQTT Alarm panels applications.   In addition to the additional states such as arming night and alarm bypass, I would like to request notifications.  

Notifications would be a new topic we subscribe to that sends us status notifications about the alarm state or possibly errors. For example, we might subscribe to topic "home/alarm/notifications" and receive the status of the alarm between states, like going from `disarmed` to `armed_home`, might send out a new notification with payload `arming_home` before we enter the new state.   

This tells us that the alarm is arming but also that it is arming to a specific state, it also is more specific than just sending `arming` state.  Having more specific information tells the UI what is happening so the user can receive proper feedback.  If I am arming away, I show an animation that the user must exist before the alarm is armed.  If I only know the "arming" state, I do not know that it is arming home, bypass, night, or away. 

We might also be able to send a notification if the user send an incorrect code to arm or disarm the alarm by sending back a notification message with payload "invalid_code".   The notification topic can be expanded to include whatever additional states or information you want without having to divert from the MQTT Manual component behavior.

The notifications allows me to provide feedback on the tablet to users based on the payload.  They might get a notification for invalid code, or that a sensor is open and the alarm cannot be armed.  Notifications would more closely mirror the functionality in the Alarmo front end. 

I would be glad to help along with one of my contributors @derekatkins to add this feature to your integration.  

Thanks!

- Mister

**Comment by derekatkins** [comment]
Hi.  As @thanksmister said, I was working on a PR for the `manual_mqtt` integration to add these features, but the HA Core team seems uninterested in improving it.  Then we were pointed to Alarmo, which seems to have 95% of the features that @thanksmister and I were were looking to add!  So instead of competing, we would like to help you add that missing 5%!  I am willing to take the work I did for the other component and supply a PR to Alarmo, if you are willing to let us bring those few missing features in?

**Comment by nielsfaber** [comment]
Hello @thanksmister  and @derekatkins,
I think it could work out brilliantly, combining Alarmo backend with your frontend!
I wouldn't consider this as competing, but joining forces.
In fact your project already came up a couple of times, i think they _should_ work together already.
Alarmo currently features an (optional) MQTT interface to send out state and receive external commands. 
There are plans to have this extended:

1. Allow Alarmo to interface with an external alarm system (so adding a command output topic, and state input topic, so mirrored version of what's currently in place) see https://github.com/nielsfaber/alarmo/issues/35
2. Add extra topics for diagnostics info (like which sensor triggered, which sensor blocked arming, false pin code entered, etc.)
3. Currently I am working out the support for multiple area's. Which means multiple alarm_control_panel entities (with an optional 'master')
4. Allow a JSON mapping from external MQTT interface to the alarmo MQTT interface (so which payload structure + strings are used for which commands, how is pin code sent, etc.)

Now, to get on topic: Alarmo exposes the `arm_mode` attribute for the `alarm_control_panel` to distinguish between `armed_away`, `armed_night` etc. This attribute is exposed for all states except `disarmed`.
For me it's a possibility to add a topic for this to have this exposed over MQTT as well (the notification topic as you called it).

Maybe there's an alternative approach that's worth discussing as well:

- Alarmo will add a `command_out` topic, on which the request that was received (either by `command_in` topic, or by arming from HA) is published (e.g. `arm_home`) -> this is part of TODO item 1.
- The `state_out` topic, which is now in place, may have payload `arming` (unchanged)
- Combining these items, would give you all the info you need, in a more versatile way (versatile in the sense that HA uses input/output topics for almost all their MQTT components, it seems to interface well with 3rd party systems)

Anyway, lets consider this a kickoff for more discussion.
I'm very much open for contributing PRs, but I would prefer to discuss/align functionality before starting on the implementation part.
Currently I have quite some work locally which will also affects the MQTT implementation, i will commit these changes soon.

**Comment by derekatkins** [comment]
@nielsfaber -
> I'm very much open for contributing PRs, but I would prefer to discuss/align functionality before starting on the implementation part.

I promise I wont start coding (or migrating the code I did for `manual_mqtt`) until we agree on an approach.  Besides, it will take me time to come up to speed with alarmo, so in some cases it may just be easier to convince you to implement it :)   I was just offering to help.

I think the `command_out` approach works for differentiating between arming_home and arming_away, but we still need a notification feedback for "invalid code".  I guess that could be your TODO # 2?

I don't understand the actual use case for multiple areas?  Could you explain that a bit more?  I DO understand the need for multiple panels (e.g. I have one near my garage entry door and another in the master bedroom), but it's all one house with a coherent alarm, and only a single `alarm_control_panel` entity in HA.

In terms of the JSON mapping, the main difference I see right now is that I chose to use "action" where you chose "command" in the input json for the `command_in` MQTT functionality.  But I think that would be easy enough for @thanksmister to change in his app to align with what you've done.

**Comment by thanksmister** [comment]
I am up for whatever changes @derekatkins thinks are needed to make.  I just want to be consistent with Manual MQTT because I have many users using this without Alarmo integration and any changes would also need to be easy enough for other users to integrate using automations or whatever equivalent on other home automation platforms.   My objective with MQTT Alarm Panel is to be platform independent by leaning more on MQTT and automations to accomplish everything.

**Comment by thanksmister** [comment]
I also have a discord channel for alarm panel discussions if it makes it easier to communicate https://discord.com/invite/euh9J2d.

**Comment by nielsfaber** [comment]
> I promise I wont start coding (or migrating the code I did for manual_mqtt) until we agree on an approach. Besides, it will take me time to come up to speed with alarmo, so in some cases it may just be easier to convince you to implement it :) I was just offering to help.

I'm fine with both ways. I'm not a hero with python, but the current MQTT implementation is only like 10 lines, so easy to extend.

> I think the command_out approach works for differentiating between arming_home and arming_away, but we still need a notification feedback for "invalid code". I guess that could be your TODO # 2?

Yes, exactly. In the Alarmo context, a `notification` is literally a push/speech message. I think it will be a confusing term to use for this. Also Alarmo defines `events`, which may be `failed_to_arm` (due to a wrong code or blocking sensors). 
I have a preference to keep such data separated from the `arm_mode`. The addition of an `event` output topic (or how we should call it) is IMO a no-brainer. The question is more _when_ it will be added.
If you're OK with combining `command_out` and `state_out` data, this would be the ideal match with the planned improvements.

> I don't understand the actual use case for multiple areas? Could you explain that a bit more? I DO understand the need for multiple panels (e.g. I have one near my garage entry door and another in the master bedroom), but it's all one house with a coherent alarm, and only a single alarm_control_panel entity in HA.

Alarmo defines entry/exit delays in a centralized way, in the state machine, only customizable per `arm_mode`.
Users can only define whether a sensor should have a delayed response or immediate response.
I received feedback from users that they would like to have more control. Entering the house through the garage takes several minutes, while opening the frontdoor and disarming is a matter of 30 seconds.
As a solution to this I am adding the option to define a separate area for the house and garage, each with its own timings.
In my own use-case the garage is completely separated from the house, so it can even remain armed while the house is not.
In fact it will result in multiple `alarm_control_panel` areas (as each can have its own state), but there will be a 'master' entity which will control all simultaneously.
Long story short: the multiple areas feature offers more granularity. Things remain backwards compatible and people don't have to use it.

> In terms of the JSON mapping, the main difference I see right now is that I chose to use "action" where you chose "command" in the input json for the command_in MQTT functionality. But I think that would be easy enough for @thanksmister to change in his app to align with what you've done.

Conflicts are inevitable here. Maybe in this case the MQTT Alarm Panel can adapt to Alarmo, but in other cases it is not possible or would require changing embedded code etc. IMO its a must-have to keep this flexible on the HA-side.
Currently users can already customize the payloads for both states and commands.
For the `command_in` topic there is also some [magic](https://github.com/nielsfaber/alarmo/blob/main/custom_components/alarmo/alarm_control_panel.py#L532) in place to parse different JSON structures. Looks like `action` can be kept as it is.
With adding the `command_out` topic, the problem arises again, so I would like to try providing a once-and-for-all solution using a JSON mapping option.

Food for thought: would it be desirable to have Alarmo expose the configured arm_modes + delay settings? It seems they are currently hard-coded (correct me if I'm wrong) in MQTT Alarm Panel, but it would be super user-friendly if things can be in sync.

> I just want to be consistent with Manual MQTT because I have many users using this without Alarmo integration and any changes would also need to be easy enough for other users to integrate using automations or whatever equivalent on other home automation platforms. My objective with MQTT Alarm Panel is to be platform independent by leaning more on MQTT and automations to accomplish everything.

I understand your point perfectly.
On the other hand, things like diagnostics feedback will improve the user friendliness which they won't have with the `manual_mqtt` panel (which is intended for basic/generic functionality only).
I'm only offering to help with integrating Alarmo with your project, I see quite some room to add functionality to make the user experience more pleasant.
Since I heard multiple users about the MQTT Alarm Panel, maybe a button can be added to the Alarmo GUI for applying a set of configurations for it, that would make it a plug-and-play experience.
The goal of Alarmo is to have everything set up by GUI, so I don't want to have users bother with topics and payload stuff in YAML. This cannot be said for the `manual_mqtt` integration.
The only leap people have to make is setting up the custom_component. Luckily this also can be done without coding skills.
Also, someone also made a [great video tutorial](https://www.youtube.com/watch?v=Etm_8PNY1JU) about Alarmo.
And this project is quite new, still finding ways to simplify stuff..

**Comment by derekatkins** [comment]
> If you're OK with combining command_out and state_out data, this would be the ideal match with the planned improvements.

I'm not sure what you mean here?  Do you mean combining into one topic?  Or combining into one changeset?

> Food for thought: would it be desirable to have Alarmo expose the configured arm_modes + delay settings? It seems they are currently hard-coded (correct me if I'm wrong) in MQTT Alarm Panel, but it would be super user-friendly if things can be in sync.

Amusingly, @thanksmister  and I did this (we called it a "configuration_topic").  HA published a JSON object that contains all the HA configuration, so all the MQTT Alarm Panel needed was the MQTT server and command topic (which defaulted to the default topic), and it could self-configure based on the data received (e.g. other topics, delay times, etc).

**Comment by thanksmister** [comment]
@nielsfaber @derekatkins  it would be good to expose the sensors used (topics and active state), that part I have in the alarm panel already for displaying open/closed states of sensors.  

However, for the UI to make use of sensors, we have to know which state is the "on" or active state.  Sensor states are arbitrary as they can be on/off,yes/not,closed,opened,away,home, or anything the user wants.

I don't use delay or pending times from the configuration, instead I rely on state changes or some indicator for state changing.   I already or will support the default modes for MQTT Manual component which I think are pretty universal, so knowing these won't matter.

**Comment by nielsfaber** [comment]
@derekatkins 
> > If you're OK with combining command_out and state_out data, this would be the ideal match with the planned improvements.

> I'm not sure what you mean here? Do you mean combining into one topic? Or combining into one changeset?

I mean my proposal is to combine the data of the state topic +  command topic for deriving the `arm_mode`. 
It seems to serve more use-cases than the originally proposed `notification` topic.
For the diagnostics status info a dedicated topic could be added. 

@thanksmister 
> it would be good to expose the sensors used (topics and active state), that part I have in the alarm panel already for displaying open/closed states of sensors.

I'm not sure what you mean here.
How is your MQTT Alarm Panel currently connected to HA for syncing sensor state info?
If it's already working, why would something extra be needed?
Adding this responsibility/functionality to Alarmo seems a bit contradictive to your statement of keeping things compatible to `manual_mqtt`, or I misunderstand you here.
I'm not sure if exposing sensor states fits with Alarmo. 
Currently Alarmo only watches sensor entities when it is armed (and only the sensors which are used in that arm mode).

**Comment by thanksmister** [comment]
User enters sensor topic and the active "state" into the application.  They need to setup an automation that published the sensor state changes on that topic.   If the state is "active" sensors show green, if state is anything else, sensors will show red. The active state is the state you want for arming the alarm. Here is a screenshot of the active states, user should be able to arm the alarm.  

![panel_sensors](https://user-images.githubusercontent.com/142340/104364915-0c4d1b80-54f6-11eb-8564-aebfa70cb691.png)

On the automation side, we would want an automation that runs through the sensors looking for open states, and does not arm the alarm, instead sends notification (like alert dialog) back to device saying whatever user wants.  The user currently has to setup these automations.

**Comment by nielsfaber** [comment]
Now that the multiple-areas-feature is in place, I could pick this up soon.
To summarize, my proposal:

1. Alarmo will get a command output topic. 
2. The topic will be user-configurable.
3. Possible (default) payloads on this topic (depends on user configuration) will be:
```
DISARM
ARM_AWAY
ARM_HOME
ARM_NIGHT
ARM_CUSTOM_BYPASS
```
4. Users can customize the payloads (this is already in place for the command input topic and will be shared), i.e. `ARM_AWAY` -> `LEAVE`, `ARM_HOME` -> `STAY` etc. 
5. I'm investigating the option for supporting a JSON structure for the payload, such as:
```
{
  "command": "<command>",
  "code": "<pin or password>",
  "area": "<area_name>"
}
```
This should remain optional for third party applications that require certain format, so will be user-configurable as well (especially the keys of the JSON struct)

6. Payload will be published together/simultaneously with the state update (unless the arming fails due to open sensors). Note that alarmo supports switching between arm modes, i.e. transition `ARMED_AWAY` -> `ARMED_HOME` may occur.
7. There will be an option for retaining the payload. It may help with restoring initial state after power loss.

8.  Alarmo will get an event output topic.
9. The topic will be user-configurable.
10. This topic will contain info why a command failed (upon sending a command on the input topic):

This may be any of the following reasons:
- State transition is not allowed (e.g. requesting `ARM_AWAY` while in `TRIGGERED`)
- No code was provided while the operation requires a code.
- Wrong code or wrong code format provided.
- (future use) Too many wrong codes provided, system is temporarily locked.
- Code may not be used for the operation (code privilege level insufficient)
- One or more sensors are blocking the arming

11. This topic will contain info why the alarm triggered (similar to "One or more sensors are blocking the arming"). Usually there is only 1 sensor causing the triggering, subsequent sensor changes are ignored and thus not sent.
12. I'm considering to send data in JSON structure, like:
```
{
  "event": "<error code / identifier>",
  "message": "<string explaining the error/event>",
  "context": "<array of sensors (entity_IDs? friendly names?) involved in the event if applicable>"
}
```
I am wondering if this needs to be user-configurable, since it is only diagnostical info from Alarmo.
13. I don't think it makes sense to retain data in the event topic.

I would appreciate your feedback on these items (before I start coding 🙂)

PS: Regarding
> it would be good to expose the sensors used (topics and active state), that part I have in the alarm panel already for displaying open/closed states of sensors.

I don't think this fits well in Alarmo, as its really not meant for bridging HA entities to 3rd party apps.
I think it would be better to provide a blueprint/script for it, that users can install in HA directly. There is also a [statestream](https://www.home-assistant.io/integrations/mqtt_statestream/)component that does this (sort of).

**Comment by thanksmister** [comment]
I agree, blueprints would be easier for adding sensors to 3rd party alarms, or even if we sent out a sensors config mqtt message, but this can be outside alarmo.  

My issues with this part: 

```{
  "event": "<error code / identifier>",
  "message": "<string explaining the error/event>",
  "context": "<array of sensors (entity_IDs? friendly names?) involved in the event if applicable>"
}
```

It's not generic, can't be localized, some 3rd party apps like mine support 9 languages.  If user was able to customize the text, then that would put a lot of setup time on the user.    

Maybe you can use codes for different types of issues, but having text breaks localization.  3rd party app can ignore message/context.

You don't want to send out sensor data, but you want to send array of sensor ids or friendly names in the event?  Maybe friendly names are better, not sure entity id's would be much value to the user if they were not the ones creating the sensors (like a spouse or house guest). 

> Users can customize the payloads (this is already in place for the command input topic and will be shared), i.e. ARM_AWAY -> LEAVE, ARM_HOME -> STAY etc.

Regarding customization, this requires too much work for integration, breaks standardization with other alarm systems, breaks with HA components.

```
{
  "command": "<command>",
  "code": "<pin or password>",
  "area": "<area_name>"
}
```

I would recommend to use JSON payload or use additional states such as ARMING_HOME, ARMING_AWAY, but not have both options.   I might suggest adding to the JSON delay time before alarm activates.   Knowing that we are arming to away with a 60 second delay for example.  

I don't have any more comments outside of those.

**Comment by thanksmister** [comment]
```
State transition is not allowed (e.g. requesting ARM_AWAY while in TRIGGERED)
No code was provided while the operation requires a code.
Wrong code or wrong code format provided.
(future use) Too many wrong codes provided, system is temporarily locked.
Code may not be used for the operation (code privilege level insufficient)
One or more sensors are blocking the arming
```

I would assign an error code to each of these and pass that in the alarm event, then 3rd party apps can display localized messages to the user.

**Comment by nielsfaber** [comment]
@thanksmister 
> Maybe you can use codes for different types of issues, but having text breaks localization. 3rd party app can ignore message/context.

Agreed that an error code is needed for each condition, this is what I meant with `"event": "<error code / identifier>",`.
The`message` part could be ignored completely if the error code is interpreted on the other side. Maybe it could be used as a fallback in case the error code is unknown (I guess this is fairly common practise?).
The downside is that you will have to add mapping for these error codes to human-readable texts for display.
This is OK for me, but I understood that you don't want to be tied to Alarmo.
If you have preferences for certain codes/code format, let me know, else I will define them later on.

> You don't want to send out sensor data, but you want to send array of sensor ids or friendly names in the event? Maybe friendly names are better, not sure entity id's would be much value to the user if they were not the ones creating the sensors (like a spouse or house guest).

Yes, my proposal is that the `context` attribute will list the sensors involved in the `failed to arm` or `triggered` scenarios.
I don't know enough of your Android Panel to know whether friendly name or entity ID would be preferred.
I assumed you already keep a list of sensor entity IDs with their states from HA (this is what that blueprint MQTT bridge thing should do), so entity ID is a better identifier as it is guaranteed to be unique.
Just so you know: in Alarmo users can assign their own names to sensors which will override the HA friendly_name (i.e. the sensor name could be different to HA entity name).
I'm open to both options, just thought the entity ID would be unambiguous.

> Regarding customization, this requires too much work for integration, breaks standardization with other alarm systems, breaks with HA components.

This is needed for allowing users to combine Alarmo with other systems. I heard some people have an MQTT-enabled alarm system and they don't have the freedom to make changes on that end. 
So I want to give users the freedom to format a (JSON) payload to a way that is compatible for them.
If your Android panel is able to parse the default HA alarm commands (`DISARM`, `ARM_AWAY`, etc.) all should work fine out-of-the-box.
By default the topic will not sent JSON payload and `code` / `area` info is not exposed either.

> I would recommend to use JSON payload or use additional states such as ARMING_HOME, ARMING_AWAY, but not have both options. I might suggest adding to the JSON delay time before alarm activates. Knowing that we are arming to away with a 60 second delay for example.

As said before, I prefer the addition of the `command output` topic over the introduction of extra states (Alarmo does not have state `ARMING_HOME`).
If it is preferred for you, we can also discuss the option of sending JSON payload on the `state` topic.
The JSON struct could add more info besides the state, such as the `arm_mode` (`armed_away`, `armed_home`, etc.).
In this case reading the `command output` topic will not be needed anymore for you (I guess).
Delay time (to next state) could also be sent. 
BUT: users can (optionally) configure Alarmo to transition from arming -> armed as soon as they close their front door.
In this case the delay time would not be leading, it would be more like a _timeout_ for the arming.

**Comment by thanksmister** [comment]
I would only suggest this for the command:

```
{
  "command": "<command>",
  "code": "<pin or password>",
  "area": "<area_name>"
  "delay": <int> 0....infinity
}
```

I think the command as JSON payload is fine, I would not change the state to a JSON payload.  MQTT app users can subscribe to the command out topic and the event topic, using the data they want from these.   

I don't have code number suggestions, can be values also like event_error, event_info, or whatever, just something that points back to the exact error codes.   This way MQTT app developers can associate error code/identity with localized messages. 

Regarding sending friendly name, if the event is for display purposes, friendly name of component would be preferred as this is the localized name the user has given to the component in most cases. Entity id is of no value that I can see, MQTT app devs have no reference to the entity_id since you are not publishing that as part of a config or anything.   Send both then, if you can't decide, extra information can be ignored.

**Comment by nielsfaber** [comment]
@thanksmister the transition delay is a property that does not belong to the command, but with the state.
E.g. transitioning from `ARMED_AWAY` -> `TRIGGERED` goes via `PENDING` with a delay `entry_time`.
The command topic will not be involved in this (only in arming/disarming operation).
It seems like you're asking for having the `state` topic payload extended with this extra info.

Sending the friendly name of sensor entities I the event context information is OK for me.

**Comment by thanksmister** [comment]
The `PENDING` with a delay if not much value.  This is one thing we tried to fix with the PR @derekatkins made, we wanted to have a state between ARMED_AWAY and TRIGGERED, DISARMED  and ARM_AWAY. Yeah, this is pending, but it means nothing more than waiting.  

We were going to add new states, but you said you were going to use commands instead.  So what is the purpose of the command?  Should it not be fore telling the alarm system what is happening, like we are arming away and we are pending 60 seconds before next command?

**Comment by nielsfaber** [comment]
I was guessing that your wall panel shows a countdown timer when the system is armed and a sensor is tripped, to notify that the user has `X` time left to disarm the system 'gracefully' until the sirens start.
The question is how you could determine this `X` if it would not be sent by Alarmo.
I wish I had an Android tablet laying around, I would have tried your project myself 😋 (I sold my soul to Google Home unfortunately).

My _idea_ of the `command_out` topic, is just to relay an incoming command (arm/disarm) to MQTT, such that all devices would get notified of this. So if someone uses the Lovelace interface to arm the system, your wall panel would be notified of this.
The fact that this info helps to figure out what state comes after `ARMING` is an additional advantage, but it was not the point.
I sense that we are misunderstanding eachother here, if you think the `command_out` topic will also be useful for extra sending information, such as delay times.
I'm guessing that the ideal solution for you would be if the `state` topic would send payloads like:
```
{
  "state": "ARMING",
  "arm_mode": "ARMED_AWAY",
  "delay": 30,
}
```
It would give you all the context you need for display purposes. The `command_out` topic will not be needed for you anymore.
Perhaps @derekatkins could join to share his view on this?

**Comment by derekatkins** [comment]
>  The question is how you could determine this X if it would not be sent by Alarmo.

Right now that delay time is "hard coded" in the app (technically it's an app setting, so when you set up the app you set the delay time(s), but IMHO it would be much much better for HA/Alarmo to tell the app the delay time).

Extending the `state` topic to JSON would solve this, except for the fact that it would break all existing MQTT users, and would be a significant change from, e.g., `manual_mqtt`. 

The other option is to create a new `event` topic, which we need anyways for feedback like "Invalid code" on disarm.  Instead of `command_out` (which, technically, the panel doesn't need), or even your JSON state topic, we could put all that information into the `event` topic:
```
{
   "event": "TRIGGERED",
   "delay": 30,
}
```
While the delay times going from `DISARMED` to `ARMED_*` are generally constant, I can certainly see use-cases where there are different delay times during `TRIGGERED`.  For example, I can image a use-case where the delay on my front door is much larger than my garage door, and the delay on my basement door is 0.   So I do think sending the delay time is important (because the app can't know which sensor triggered the alarm).

**Comment by nielsfaber** [comment]
@derekatkins That's a quick reply 👍 

I agree 100% on the fact that is best to keep configuration in 1 place, in this case Alarmo, as it manages the alarm state machine.
The delay setting can definitely not be represented by a constant, for Alarmo this depends on which area a sensor is assigned to (garage vs hallway as you said) and a sensor may also be configured to have zero delay (tamper or fire sensor).

So for the Android MQTT Panel, it is preferred to leave the `state` topic untouched, so only send state info here. 
All contextual information can be shared through the `event` topic.

So far I only considered non-happy-flow `events` to be sent here (blocking sensors, wrong pincode etc).
But it seems a good idea to extend this and also send payloads during other state transitions, which will contain the contextual info you are looking for.
When arming, I could send as event `ARM_AWAY` together with the exit delay. 
When triggered, the payload would be as you said.

Let's do it like this, I guess we can forget about the `command_out` topic then (it seems not needed anymore).

Agreed?

**Comment by derekatkins** [comment]
I should add, I think my suggestion above is missing data.  I think it should be:
```
{
   "event": "TRIGGERED",
   "delay": 30,
   "sensor": [ .... ]
}
```
(and maybe additional information?)

**Comment by thanksmister** [comment]
> I should add, I think my suggestion above is missing data. I think it should be:
> 
> ```
> {
>    "event": "TRIGGERED",
>    "delay": 30,
>    "sensor": [ .... ]
> }
> ```
> 
> (and maybe additional information?)

What will be the all the event names?  ARMING_HOME? Would be nice to now the between states from disarmed to armed home, should be arming home with delay time?

**Comment by nielsfaber** [comment]
HA [defines](https://github.com/home-assistant/core/blob/dev/homeassistant/components/mqtt/alarm_control_panel.py#L68) the following:
```
ARM_AWAY
ARM_HOME
ARM_NIGHT
ARM_CUSTOM_BYPASS <--- don't know if you support this one
DISARM
```
I'm OK with adding the `-ing` behind all, no preference.

For the initial triggering (so from `ARMED_AWAY` -> `PENDING`), I now (internally) use `ENTRY` (the delay is also called entry delay in Alarmo).
I'm also OK with `TRIGGERING` but technically this comes later (when the sirens start). 
Note that a sensor may also have zero entry delay, in this case it would skip the `ENTRY` event, it would have event `TRIGGER(-ING?)`.
Note that Alarmo currently will only store+send the sensor which was the root cause of the triggering, not the subsequent tripped sensors.
I don't know if this gives any issues for you?

For the error events we could use the following:
```
TRANSITION_NOT_ALLOWED: State transition is not allowed (e.g. requesting ARM_AWAY while in TRIGGERED)
CODE_MISSING: No code was provided while the operation requires a code.
CODE_INVALID: Wrong code or wrong code format provided.
CODE_LOCKOUT: (future use) Too many wrong codes provided, system is temporarily locked.
CODE_NOT_ALLOWED: Code may not be used for the operation (code privilege level insufficient)
FAILED_TO_ARM: One or more sensors are blocking the arming
```

Just some proposal, please shoot 🙂

**Comment by nielsfaber** [comment]
I added this functionality in the v1.5.2 release of Alarmo.
I wrote down the current implementation in the [documentation](https://github.com/nielsfaber/alarmo#event-topic).
Note that the naming of the events has been changed a bit compared my proposal above.

Couple of things have been altered due to change of insight while working on it:
* When a sensor triggers the alarm, the state transition `ARMED_<mode>`->`PENDING`->`TRIGGERED` occurs.
I decided to only fire the `TRIGGER` event on the first transition.
The `delay` attribute will tell you when the `PENDING`->`TRIGGERED` state transition will follow (there is no need for firing an extra event). `delay` may be zero, in this case the `PENDING` state will be skipped.
* When arming, the state transition `DISARMED`->`ARMING`->`ARMED_xx` occurs.
I decided to only fire the `ARM_<mode>` on the first transition (so same as for triggering).
The `delay` attribute will tell you when the `ARMING`->`ARMED_<mode>` state transition will follow (there is no need for firing an extra event). `delay` may be zero, in this case the `ARMING` state will be skipped.
* `CODE_INVALID` and `CODE_NOT_ALLOWED` are merged into `INVALID_CODE_PROVIDED`. It's rather difficult to make distinction in the code, it doesn't give much extra info anyway.
* `CODE_LOCKOUT` is not implemented yet (since Alarmo does not have any lockout functionality yet). The event list may grow anyway, so perhaps its good to add some catch for unrecognized events.
* I don't publish event messages (yet). The event names should be(come) self-explanatory.

I hope the documentation provided a clear picture of what's in place.
There's no problem if things need to be changed or added, just let me know.

**Comment by nielsfaber** [comment]
@thanksmister @derekatkins Any comments on this?
It feels a bit weird to me that the discussion suddenly stopped, after spending effort adapting things to your needs.
Or are you busy adapting things on your side 😉 ?

**Comment by derekatkins** [comment]
Sorry, I was letting @thanksmister do this thing (they are the App developer; I was working on the `manual_mqtt` update before I found Alarmo).  I think it sounds good.

**Comment by thanksmister** [comment]
I am currently away on vacation, but I think the latest rounds of changes work for me fine.  I will leave any details upt to @derekatkins as he knows best the `manual_mqtt` component, thanks!

**Comment by nielsfaber** [comment]
Fair enough, let me know if things are unclear or need to be changed.

**Comment by thanksmister** [comment]
@nielsfaber can you help define all the new events.  The documentation doesn't seem to list the new ones.

**Comment by nielsfaber** [comment]
The documentation is up to date with the current implementation.
What are you missing?

**Comment by thanksmister** [comment]
> * When a sensor triggers the alarm, the state transition `ARMED_<mode>`->`PENDING`->`TRIGGERED` occurs.
>   I decided to only fire the `TRIGGER` event on the first transition.
>   The `delay` attribute will tell you when the `PENDING`->`TRIGGERED` state transition will follow (there is no need for firing an extra event). `delay` may be zero, in this case the `PENDING` state will be skipped.
> * When arming, the state transition `DISARMED`->`ARMING`->`ARMED_xx` occurs.
>   I decided to only fire the `ARM_<mode>` on the first transition (so same as for triggering).
>   The `delay` attribute will tell you when the `ARMING`->`ARMED_<mode>` state transition will follow (there is no need for firing an extra event). `delay` may be zero, in this case the `ARMING` state will be skipped.

Hi, I guess I expected to have the events and or commands in the documentation updated with examples of the arming, disarming, pending states?  We had talked about using commands and or/events with delay times when transitioning between states of the alarm.  I am not sure now if we were going to rely on commands or events, I think we were going to update commands?     @derekatkins maybe you can help sort this out, it's a long thread.   I think the commands were to have the arming_home, arming_away, arming_night states as well as the delay times for each?

**Comment by thanksmister** [comment]
According to the documentation, commands are sent only for controlling alarmo:

> Command topic
> The command topic can be used for external control of Alarmo through MQTT.

And for events we have:

> The event topic is used for publishing additional status updates of the alarm_control_panel entities.

But from this thread, it seems that you intended to send out "commands" from alarmo:

`As said before, I prefer the addition of the command output topic over the introduction of extra states (Alarmo does not have state ARMING_HOME).`

I think we need to have either the command or event feature further defined with the extra states in the documentation.  I also think now we have kind of an issue with event and commands having the same functionality.   From the application, I send commands and listen to events.   Events would be where I would expect the ARMING_HOME state to live, along with the delay time.

**Comment by derekatkins** [comment]
I'll take a look.  My understanding, @thanksmister , is that there are three topics, the status topic, the command topic and the event topic.
* The status topic gives you the current state (`DISARMED`, `ARMING`, `ARMED_xxx`, `PENDING`, `TRIGGERED`, etc).
* The command topic is input, where you send MQTT commands to control alarmo (JSON with command and code), e.g. `{ "command": "disarm", "code": "1234" }`
* The event topic is for extra information during (some) state transitions.  You wont get an event for ALL state transitions.  For example, it will give you the `ARMING_HOME` event with a delay, or `TRIGGERED` with a delay.  It won't give you another event goes into `ARMED_HOME` or `TRIGGERED`, but that's okay because you'll get a status message when that transition occurs.  The event message is supposed to be JSON, but I'm not 100% sure of the format (I have to go look at the docs, but wont have time to do so for several hours).

**Comment by thanksmister** [comment]
> * The status topic gives you the current state (`DISARMED`, `ARMING`, `ARMED_xxx`, `PENDING`, `TRIGGERED`, etc).

I think you mean the state topic? The one used by mqtt manual called `state_topic`

**Comment by nielsfaber** [comment]
@thanksmister I think you missed this post: https://github.com/nielsfaber/alarmo/issues/92#issuecomment-769927761
From our communication I understood that it was preferred for you to have the event topic added, which tells you additional info related to state transitions.
It is exactly made as @derekatkins describes:
- When the user sets Alarmo for arming, an `ARM_AWAY` message is fired in the event topic, with the delay until the alarm is armed (so until transition `ARMING->ARMED_AWAY`) as attribute.
- Similarly, if a sensor triggers the alarm, a `TRIGGER` message is fired, with the delay as attribute (between transition `PENDING->TRIGGERED`) and the sensor that caused it as attribute.

Please read the docs on the exact payload format, or just install Alarmo and play around with it.
I'm quite sure all is in place that was requested.

**Comment by thanksmister** [comment]
@nielsfaber I have noticed what I think is a race condition. When setting the alarm to home mode from HA interface, I sometimes get the event arm_home after the state armed_home.   This makes it difficult to determine which state to follow.  I am not sure having both events and states for alarm states is best approach.   Would maybe have been better if we just modified alarm states with a JSON payload and used events for errors.   Not sure how to address it without just a hack to delay state after events.   Thanks!

**Comment by nielsfaber** [comment]
@thanksmister The event and state updates are handled asynchronously so there is indeed no guarantee that one comes before the other. 
In mqtt there is (AFAIK) also no guarantee that messages arrive at the target in the order in which they are published anyway, so I doubt it can be fixed completely.
In fact not all state transitions comewith  an event, and the other way around, an event can come without a state update.
I think your implementation should take this into account.

I don't want to merge the state and event data into a single topic, as the event data is only providing contextual info so it has a different function. 
The state payloads can also not become JSON formatted, as this will (irreversibly) break things for existing users.
Only listening to the state topic will not be a solution either, because as said, not all state updates come with an event (like disarming).

Can you clarify why this behaviour is a problem for you?
I was assuming the arm delay would just help you with showing a countdown timer in the GUI, but if the delay is zero this can be skipped right?

**Comment by derekatkins** [comment]
Perhaps if delay=0 we skip sending the event?  Or if delay=0 you ignore the event?  I don't know which is better/easier.

**Comment by thanksmister** [comment]
Yeah for my app, I read in the event/state and set the "state" of the alarm.  Originally I think we intended to have new states, rolling them over into events means now we have states and events with alarm mode/state values and then this issue.  

I parse out the event and save the state of the alarm to the database, same with states, then I update the UI based on the latest state in the database.   So I would have to hack this to get it to work.  

I still think we made a possible bad design choice by adding states or alarm modes to events.  We should have kept states for mqtt manual and then moved all states to events so end user can choose to use events or states, but would not use both.  I am not suggesting you do that, just that this seems to be the issue.

**Comment by derekatkins** [comment]
Let me rephrase what @thanksmister is saying:  I think the `status` topic should be `state++`.  In other words, I think Alarmo needs to duplicate every message sent to `state` also to `status`, where `state` is just the raw state and `status` is JSON, so contains all the other information we want/need.  

A listener should only subscribe to one topic and be able to get all the data they want/need.  I.e., they should not need to subscribe to both topics, the `status` topic should be sufficient and get all messages sent.

**Comment by nielsfaber** [comment]
@thanksmister @derekatkins 

> A listener should only subscribe to one topic and be able to get all the data they want/need. I.e., they should not need to subscribe to both topics, the events topic should be sufficient and get all messages sent.

I think this is a matter of preference.
If you take a look for example at HA's [mqtt light](https://www.home-assistant.io/integrations/light.mqtt/) component, you see that it has separate `state`, `brightness`, `color`, `effect` etc. input/output topics.
Similar things can be seen throughout all HA's mqtt components.

With Alarmo's `state`/`command` topics I tried to stick to the same structure, this should be inline with the `manual_mqtt` component as well.
The `event` topic is another thing, as it combines different data in a single channel, and it has JSON format.
My idea was to have Alarmo only publish "contextual" data here, giving extra info about a state transition that occured (delay time, involved sensor(s), etc.) or a reason why a arm/disarm command was not allowed.
I thought we both agreed on this in https://github.com/nielsfaber/alarmo/issues/92#issuecomment-769927761.

Apparently there is some misconception here, since you seem to think that the `event` topic provides a substitute for the `state` topic.
In its current implementation it's not designed like this, for two reasons:
* There are less events defined than there are state transitions (see [here](https://github.com/nielsfaber/alarmo#event-topic) for the list of events)
* Events are not only fired on state transitions, but also provide feedback when handling commands (wrong code provided, etc.)

I'm a bit hesitant to shuffling things up yet again, given the fact we sent like 20 messages back and forth here before I started implementing what's now in place (which was meant to fulfil your feature request).
It gives me the feeling that later in time you may change insights yet again, and Alarmo needs to keep adapting.
Alarmo is not designed specifically as a back-end for your project, it serves a wider audience.
I would like to keep things generic as much as possible, which means that I want to avoid that the `event` topic becomes specific for your use-case only, and becomes a replacement for the `state` topic.
I was under the impression that you wanted to maintain compatibility with the `manual_mqtt` component, hence the `event` topic serves as an (optional) layer on top of the `state` topic.
I still see elegance in this format, so I would prefer to stick to it, rather than sending duplicate info.

Some questions from my side:
* Could you share some insight in why it is a problem for you to subscribe to both the `state` and `event` topic? I hope you know that mqtt allows subscribing to multiple topics at once using a wildcard (like `/alarmo/*` to catch `/alarmo/state` and`/alarmo/event`)?
* In which scenarios do you see difficulties combining the data from both topics to update the GUI? Are we only talking about cases where the alarm is armed without delay time?
* If I would be able to tackle the 'race condition' where the`state` update  comes before `events` info, would things be solved for you?
* If all looks hopeless here, could you make a proposal how you would like to see the behaviour + payloads of the `event` topic?

**Comment by thanksmister** [comment]
@nielsfaber I think we have enough at this point to allow for a good integration with alarm panel application and alarmo.  I don't think we need any additional changes, I was only suggesting how we might fix the race condition.  

However, a fix is not needed for the integration point with Alarm Panel.   I want to thank you for considering all the changes and updating the code for Alarm Panel.  I am going to close this ticket as I think its complete. 

Thanks again for all work!!

**Comment by nielsfaber** [comment]
@thanksmister Can you give an update on the status of the integration for Alarmo on your side?
Thanks!

**Comment by thanksmister** [comment]
I am working on it and there is already early versions.  Remendado that it already works with Alarmó because it works with MQTT manual.  The test version also sends the code which is compatible with alarmó. The only other integration point is listening for some event messages for errors and the state changes.

**Comment by nielsfaber** [comment]
@thanksmister OK great to hear there is progress on this!
Please keep me posted if there's any updates.

**Comment by denisjoshua** [comment]
> [thanksmister](https://github.com/thanksmister) Sorry if I ask you, but I like a lot the colors that your dashbord have, 
Can you tell me the name of your dashboard please ? 
Thanks a lot
